src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs

```diff
@@ -100,7 +97,7 @@ internal sealed class Builder
                 /// <summary>
                 /// Snapshots are kept in a dictionary of position -> snapshot at that position. These are stored in descending order.
                 /// </summary>
-                private readonly ImmutableSortedDictionary<int, Snapshot>.Builder _incrementalSnapshots = ImmutableSortedDictionary.CreateBuilder<int, Snapshot>(DescendingIntComparer.Singleton);
+                private readonly SortedDictionary<int, Snapshot> _incrementalSnapshots = new SortedDictionary<int, Snapshot>();
```

Comment:
>SortedDictionary [](start = 33, length = 16)

Do we really need to use sorted dictionary? Could we use PooledDictionary and simply sort the content of the ArrayBuilder before ```ToImmutableAndFree```?

Reply:
We could, but I assumed that this would have slightly better perf characteristics. `SortedDictionary.Add` is `O(log n)`, which will we will do `n` times, but the `log n` factor will grow slightly slower because the `n` starts smaller. Sorting, by comparison, will always be pure `O(n log n)`.

---
In reply to: [295507816](https://github.com/dotnet/roslyn/pull/36563#discussion_r295507816) [](ancestors = 295507816)

Reply:
So I decided to get some actual perf numbers on this to back up my assumptions. It looks like just using an `ArrayBuilder` here and sorting based on the tuple position is by far the most performant option, so I'll move to that.

---
In reply to: [295552408](https://github.com/dotnet/roslyn/pull/36563#discussion_r295552408) [](ancestors = 295552408,295507816)

Reply:
I decided to get some data on the difference here:

|             Method |    N |           Mean |        Error |      StdDev |
|------------------- |----- |---------------:|-------------:|------------:|
|   SortedDictionary |   10 |       977.3 ns |     40.43 ns |    119.2 ns |
| DictionaryThenSort |   10 |     1,371.7 ns |     46.47 ns |    137.0 ns |
|   SortedDictionary |   50 |     6,099.4 ns |    120.26 ns |    264.0 ns |
| DictionaryThenSort |   50 |     6,930.7 ns |    127.36 ns |    119.1 ns |
|   SortedDictionary |  100 |    10,617.2 ns |    211.68 ns |    531.1 ns |
| DictionaryThenSort |  100 |    10,881.7 ns |    217.44 ns |    344.9 ns |
|   SortedDictionary | 1000 |   182,388.9 ns |  1,849.91 ns |  1,639.9 ns |
| DictionaryThenSort | 1000 |   176,751.8 ns |  3,528.04 ns |  3,465.0 ns |
|   SortedDictionary | 5000 | 1,176,685.3 ns | 23,136.48 ns | 40,521.7 ns |
| DictionaryThenSort | 5000 | 1,070,438.2 ns | 16,443.11 ns | 15,380.9 ns |

The differences here are negligible, so unless you feel strongly I'm going to leave it as is.

---
In reply to: [295552408](https://github.com/dotnet/roslyn/pull/36563#discussion_r295552408) [](ancestors = 295552408,295507816)